File: frontend/src/services/readinessService.ts
import axiosInstance from '@/utils/axios';
import type {
  ReadinessCategory,
  ReadinessDepth,
  ReadinessReport,
  ReadinessRunList,
  ReadinessSummary,
  ReadinessTrigger,
} from '@/types/readiness';

/** Body of POST /agent/{id}/readiness. Optional fields are omitted when
 * unset so backend defaults apply. */
interface ReadinessRequestBody {
  depth: ReadinessDepth;
  config_id?: string;
  trigger?: string;
  /** Targeted-deep filter — only meaningful when depth=`deep`. Backend probes
   * only these categories live; the rest return SKIPPED. Used by save flows
   * that know exactly which resources changed. */
  categories?: ReadinessCategory[];
  /** Force a fresh run + persisted event, bypassing the backend read-through
   * caches. Set by the "Run deep test" button so each explicit run appears as a
   * new entry in the run-history list. */
  force?: boolean;
}

/** Derive a badge-shaped {@link ReadinessSummary} from a full report. Kept
 * client-side so a targeted-deep POST /readiness call can drive the same UI
 * atom the shallow GET /readiness/summary call feeds — no extra round-trip. */
export const reportToSummary = (report: ReadinessReport): ReadinessSummary => ({
  agent_id: report.agent_id,
  config_id: report.config_id,
  overall_status: report.overall_status,
  blocker_count: report.summary.blockers,
  warning_count: report.summary.warnings,
  info_count: report.summary.info,
});

/** Marker the backend runner sets as the skip reason for deep checks it
 * filtered out because the caller only requested a subset of categories.
 * Mirrors `TARGETED_DEEP_SKIP_PREFIX` in `core/services/readiness/runner.py`.
 * The coupling is enforced by `test_readiness_frontend_contract.py` — that test
 * fails if the backend value changes, so this copy can't silently drift.
 * Substring-matched to avoid coupling to the exact templated category name. */
const TARGETED_DEEP_SKIP_MARKER = 'Not re-probed on this save';

/**
 * Pairs of (shallow heads-up check-id, authoritative deep check-id). When the
 * deep check produced a real result for a resource, the shallow heads-up is
 * redundant and dropped. MIRRORS `core/services/readiness/consolidation.py`
 * (`_SUPERSEDED_BY_DEEP`) — the backend already de-duplicates fresh reports;
 * this is applied again after a targeted-deep MERGE (which can re-introduce a
 * carried-forward deep row next to a fresh shallow heads-up). The coupling is
 * enforced by `test_readiness_frontend_contract.py`, which fails if the backend
 * pairs change, so this copy can't silently drift. check-ids are matched by
 * prefix so per-resource ids match.
 */
const DEEP_SUPERSEDES_SHALLOW: ReadonlyArray<readonly [string, string]> = [
  ['mcp_servers.oauth_token_valid', 'mcp_servers.reachable'],
  ['tools.oauth_token_valid', 'tools.reachable'],
];

const matchesCheckId = (id: string, prefix: string): boolean =>
  id === prefix || id.startsWith(`${prefix}:`);

/** Drop shallow heads-up rows the deep probe has already answered. Pure. */
const suppressRedundantShallowChecks = (
  checks: ReadinessReport['checks'],
): ReadinessReport['checks'] => {
  const ranDeep = new Set<string>();
  for (const c of checks) {
    if (c.status === 'skipped') continue;
    for (const [, deep] of DEEP_SUPERSEDES_SHALLOW) {
      if (matchesCheckId(c.check_id, deep)) ranDeep.add(deep);
    }
  }
  if (ranDeep.size === 0) return checks;
  const superseded = DEEP_SUPERSEDES_SHALLOW.filter(([, deep]) => ranDeep.has(deep)).map(
    ([headsup]) => headsup,
  );
  return checks.filter((c) => !superseded.some((prefix) => matchesCheckId(c.check_id, prefix)));
};

/**
 * Merge a fresh targeted-deep report with the previous one.
 *
 * The targeted-deep flow (see `refreshReadinessAfterSave` in
 * `AgentEditorShell`) asks the backend to only live-probe the categories the
 * user just touched. Every other deep check comes back SKIPPED with a
 * dedicated marker so the report shape stays uniform. Without merging, a
 * second consecutive save with a different category would appear to "clear"
 * the previous save's warnings — e.g. attach a broken tool → save (deep
 * probes tools, FAIL surfaces), change LLM → save (deep re-probes LLM,
 * tool check comes back SKIPPED, tool FAIL vanishes from the badge).
 *
 * This merge preserves the last known real status for every check the new
 * report didn't actually re-run, then recomputes summary counts and
 * `overall_status` so the badge reflects the union.
 */
export const mergeReadinessReports = (
  prev: ReadinessReport | null,
  next: ReadinessReport,
): ReadinessReport => {
  if (!prev) return next;

  const prevByCheckId = new Map(prev.checks.map((c) => [c.check_id, c]));
  const nextIds = new Set(next.checks.map((c) => c.check_id));

  const isFilteredMarker = (c: ReadinessReport['checks'][number]): boolean =>
    c.status === 'skipped' &&
    (c.skip_reason ?? c.message ?? '').includes(TARGETED_DEEP_SKIP_MARKER);

  // Bare check_ids the new run left unprobed (targeted-deep save that didn't
  // touch that category). A per-resource check (e.g. one row per MCP server)
  // emits rows keyed `${bareId}:${resourceId}` when it DOES run, but only the
  // bare marker row when it's filtered — so we match prior per-resource rows
  // by prefix, not exact id.
  const filteredBareIds = next.checks.filter(isFilteredMarker).map((c) => c.check_id);
  const belongsToFilteredCheck = (id: string): boolean =>
    filteredBareIds.some((bare) => id === bare || id.startsWith(`${bare}:`));

  const mergedChecks: ReadinessReport['checks'] = [];
  for (const c of next.checks) {
    if (!isFilteredMarker(c)) {
      mergedChecks.push(c);
      continue;
    }
    // Stable-id check whose category was untouched: restore the prior real
    // outcome under the SAME id.
    const prior = prevByCheckId.get(c.check_id);
    if (prior && prior.status !== 'skipped') {
      mergedChecks.push(prior);
      continue;
    }
    // Per-resource check (dynamic ids): drop the bare marker only when the
    // prior report actually has per-resource rows to carry forward below —
    // otherwise keep the marker so the category isn't silently empty.
    const hasPriorResourceRows = prev.checks.some(
      (p) =>
        p.status !== 'skipped' &&
        !nextIds.has(p.check_id) &&
        p.check_id.startsWith(`${c.check_id}:`),
    );
    if (!hasPriorResourceRows) mergedChecks.push(c);
  }

  // Carry forward prior per-resource rows whose check was filtered out this
  // run and that have no counterpart in the new report. Without this, a
  // targeted-deep save that didn't touch MCP would drop previously-failing
  // per-server MCP rows from the badge until the next deep test.
  for (const p of prev.checks) {
    if (p.status !== 'skipped' && !nextIds.has(p.check_id) && belongsToFilteredCheck(p.check_id)) {
      mergedChecks.push(p);
    }
  }

  // Re-apply the backend's cross-check de-duplication: a carried-forward deep
  // row (e.g. "MCP can't be reached") next to a fresh shallow heads-up
  // ("token may be expired") for the same resource would show the same problem
  // twice. Run BEFORE counting so the badge numbers match what's rendered.
  const dedupedChecks = suppressRedundantShallowChecks(mergedChecks);

  let blockers = 0;
  let warnings = 0;
  let info = 0;
  let passed = 0;
  let skipped = 0;
  for (const c of dedupedChecks) {
    if (c.status === 'pass') passed += 1;
    else if (c.status === 'skipped') skipped += 1;
    else if (c.status === 'fail') {
      if (c.severity === 'blocker') blockers += 1;
      else if (c.severity === 'warning') warnings += 1;
      else info += 1;
    }
  }
  const overall_status: ReadinessReport['overall_status'] =
    blockers > 0 ? 'not_ready' : warnings > 0 ? 'ready_with_warnings' : 'ready';

  return {
    ...next,
    overall_status,
    summary: { blockers, warnings, info, passed, skipped },
    checks: dedupedChecks,
  };
};

/**
 * Run a readiness check for one agent + (optionally explicit) config.
 * - Depth `shallow`: DB-only, safe on hot paths.
 * - Depth `deep`: hits real providers; rate-limited to 1/min per agent server-side.
 *
 * Errors bubble up unmodified so the caller can inspect status + payload
 * (needed by the publish flow to distinguish `readiness_blocked` vs
 * `readiness_warnings`). Non-publish callers should wrap in `handleApiError`.
 */
export const getAgentReadiness = async (
  agentId: string,
  depth: ReadinessDepth,
  configId?: string,
  trigger?: ReadinessTrigger | string,
  categories?: ReadinessCategory[],
  force?: boolean,
): Promise<ReadinessReport> => {
  const body: ReadinessRequestBody = { depth };
  if (configId) body.config_id = configId;
  if (trigger) body.trigger = trigger;
  if (categories && categories.length > 0) body.categories = categories;
  if (force) body.force = true;
  const res = await axiosInstance.post<ReadinessReport>(`/agent/${agentId}/readiness`, body);
  return res.data;
};

/**
 * Cheap summary shape for the agent list badge. Always Shallow — the badge
 * doesn't need per-check detail. Trigger defaults to `list_page` if omitted
 * so backend analytics can distinguish list impressions from editor loads.
 */
export const getAgentReadinessSummary = async (
  agentId: string,
  configId?: string,
  trigger: ReadinessTrigger | string = 'list_page',
): Promise<ReadinessSummary> => {
  const res = await axiosInstance.get<ReadinessSummary>(`/agent/${agentId}/readiness/summary`, {
    params: { ...(configId ? { config_id: configId } : {}), trigger },
  });
  return res.data;
};

/**
 * List past deep readiness runs (newest first) for the run-history dropdown.
 * Metadata only — call {@link getAgentReadinessRun} to load one run's checks.
 */
export const listAgentReadinessRuns = async (
  agentId: string,
  limit = 20,
): Promise<ReadinessRunList> => {
  const res = await axiosInstance.get<ReadinessRunList>(`/agent/${agentId}/readiness/runs`, {
    params: { limit },
  });
  return res.data;
};

/** Fetch the full stored report (with checks) for one past run. */
export const getAgentReadinessRun = async (
  agentId: string,
  runNumber: number,
): Promise<ReadinessReport> => {
  const res = await axiosInstance.get<ReadinessReport>(
    `/agent/${agentId}/readiness/runs/${runNumber}`,
  );
  return res.data;
};

File: core/services/readiness/readiness_service.py
"""ReadinessService — the single class routers depend on.

Composes the pieces of the readiness domain:

- ``ContextBuilder``       — assembles a ``CheckContext`` from the DB
- ``Runner``               — executes applicable checks and aggregates
- ``DeepCheckCache``       — reuses recent Deep results & coalesces concurrent runs
- ``RateLimiter``          — caps Deep runs at 1/minute per (org, agent)
- ``ReadinessPersistence`` — writes snapshots + events to Postgres

The cache and rate limiter are module-level singletons because they hold
process-wide state; everything else is instantiated per request.

Read path also has an **edit-based fast-path**: we compute the pipeline's
existing ``dependency_stamp`` and, if a stored snapshot has the same stamp,
return it directly — never re-running the runner when nothing has changed.
"""

from __future__ import annotations

import time
from datetime import datetime, timedelta, timezone
from typing import Dict, List, Optional, Set, Union
from uuid import UUID

from fastapi import HTTPException, status
from loguru import logger

from sqlalchemy import func, select

from core.models.agent import Agent
from core.models.agent_readiness_event import AgentReadinessEvent
from core.models.agent_readiness_snapshot import AgentReadinessSnapshot
from core.models.phone_number import PhoneNumber
from core.services.base import BaseService
from core.services.pipeline.service_resolver import (
    compute_agent_cache_version,
    get_active_agent_config,
)
from core.services.readiness.cache import DeepCheckCache
from core.services.readiness.context import ContextBuilder
from core.services.readiness.persistence import ReadinessPersistence
from core.services.readiness.rate_limiter import RateLimiter
from core.services.readiness.runner import Runner
from core.services.readiness.schemas import (
    Category,
    CheckResult,
    Depth,
    OverallStatus,
    ReadinessReport,
    ReadinessRunList,
    ReadinessRunListItem,
    ReadinessSummary,
)


# Process-wide singletons — see docstring above.
_deep_cache = DeepCheckCache(ttl_seconds=300)
_rate_limiter = RateLimiter(max_per_window=1, window_seconds=60)

# Wall-clock TTL for the persisted Deep-snapshot fast-path.
#
# The ``dependency_stamp`` invalidates a stored snapshot on every DB edit
# tracked by ``compute_agent_cache_version``, but it has no way to notice
# *out-of-band* drift — an OAuth token silently expiring, a provider
# revoking an API key, a telephony account running out of credit. Without
# a time cap, the last passing Deep snapshot can be replayed for weeks
# while the agent is actually broken.
#
# 300s matches the in-memory ``DeepCheckCache`` TTL above so both cache
# layers share a single freshness window: within 5 minutes of any external
# drift the runner re-executes live, bringing the failure into the report.
_DEEP_SNAPSHOT_TTL_SECONDS = 300


# Canonical trigger strings. Free-form so callers can pass anything, but
# keeping the set here documents intent and makes analytics comparable.
TRIGGER_API = "api"
TRIGGER_PUBLISH_GATE = "publish_gate"


class ReadinessService(BaseService):
    """Public facade over the readiness subsystem."""

    async def check(
        self,
        agent_id: str,
        depth: Depth = Depth.SHALLOW,
        config_id: Optional[str] = None,
        *,
        trigger: str = TRIGGER_API,
        deep_categories: Optional[Set[Category]] = None,
        force: bool = False,
    ) -> ReadinessReport:
        """Run a readiness check for one agent + (optionally explicit) config.

        Order of operations:
        1. Load the agent + compute the dependency stamp (single-round-trip).
        2. If a stored snapshot exists with a matching stamp → return it.
        3. Shallow: run + persist + return.
        4. Deep: rate-limit, coalesce, cache-in-memory, run + persist + return.

        ``deep_categories`` is only meaningful when ``depth=DEEP`` and switches
        the runner to a **targeted deep** — only the listed categories are
        actually probed live; other deep checks return SKIPPED. Targeted deep
        deliberately bypasses the snapshot fast-path AND the persistence write
        so a partial-scan result never masquerades as a full deep for the
        publish gate. It also skips the rate limiter / coalesce cache because
        each targeted run has a different "shape" — those safeguards protect
        full-deep runs, not user-driven save-time probes.

        ``force`` is for an **explicit, user-initiated run** (the "Run deep
        test" button). It skips the read-through snapshot fast-path AND the
        in-memory coalesce cache so the run actually executes and appends a
        fresh ``agent_readiness_events`` row — otherwise a repeat deep test on
        an unchanged agent returns the stored snapshot and the run-history list
        never grows. The rate limiter still applies (1/min).
        """
        agent = self._require_agent(agent_id)
        stamp, resolved_config = self._compute_stamp(agent)
        effective_config_id = self._effective_config_id(config_id, resolved_config)

        # Targeted deep bypasses all caches + persistence — see docstring.
        if depth == Depth.DEEP and deep_categories is not None:
            return await self._run(
                agent, config_id, Depth.DEEP, deep_categories=deep_categories
            )

        # A forced run skips the read-through fast-path so it always executes
        # and records a new event (see docstring).
        if not force:
            fresh = self._read_fresh_snapshot(agent.id, effective_config_id, depth, stamp)
            if fresh is not None:
                return fresh

        if depth == Depth.SHALLOW:
            return await self._run_and_persist(
                agent=agent,
                config_id=config_id,
                depth=Depth.SHALLOW,
                trigger=trigger,
                dependency_stamp=stamp,
            )

        # Deep path: rate-limit + in-memory coalesce (a second belt over
        # the DB fast-path for the "click Test agent twice fast" case).
        cache_key = DeepCheckCache.key(self.org_id, agent.id, effective_config_id)

        async def _compute() -> ReadinessReport:
            allowed = await _rate_limiter.try_acquire(
                RateLimiter.key(self.org_id, agent.id)
            )
            if not allowed:
                raise HTTPException(
                    status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                    detail=(
                        "Deep readiness check rate-limited. "
                        "Please wait a minute and try again."
                    ),
                )
            return await self._run_and_persist(
                agent=agent,
                config_id=config_id,
                depth=Depth.DEEP,
                trigger=trigger,
                dependency_stamp=stamp,
            )

        # A forced deep run bypasses the in-memory coalesce cache too, so it
        # truly re-probes and appends a new event instead of returning a report
        # cached from a run in the last 300s. Still rate-limited via _compute.
        if force:
            return await _compute()
        return await _deep_cache.get_or_compute(cache_key, _compute)

    async def summary(
        self,
        agent_id: str,
        config_id: Optional[str] = None,
        *,
        trigger: str = TRIGGER_API,
    ) -> ReadinessSummary:
        """Small badge shape (Shallow only) — for the agent list page."""
        report = await self.check(
            agent_id, Depth.SHALLOW, config_id, trigger=trigger
        )
        return report.to_summary()

    def latest_for_agents(self, agent_ids: List[UUID]) -> Dict[UUID, Dict]:
        """Batch-fetch the latest readiness snapshot per agent for the list badge.

        Read-only: returns the most recently computed stored snapshot (any
        config / depth) verbatim — no recompute, no dependency-stamp
        re-validation. Powers the agent-list badge in a single query instead of
        one live readiness call per row (the old N+1); the editor drawer still
        recomputes accurately when opened. An agent with no stored run yet is
        simply absent from the map (badge renders "unavailable").

        Org-scoped by ``self.org_id``. Projects only the badge columns (skips the
        heavy ``checks`` blob) and reuses ``_counts_from_mapping`` so the count
        fallback set stays single-sourced with the row mappers.

        Returns ``{agent_id: {overall_status, blocker_count, warning_count,
        info_count, run_number, computed_at}}``.
        """
        if not agent_ids:
            return {}
        rows = (
            self.db.query(
                AgentReadinessSnapshot.agent_id,
                AgentReadinessSnapshot.overall_status,
                AgentReadinessSnapshot.counts,
                AgentReadinessSnapshot.run_number,
                AgentReadinessSnapshot.computed_at,
            )
            .filter(
                AgentReadinessSnapshot.organization_id == self.org_id,
                AgentReadinessSnapshot.agent_id.in_(agent_ids),
            )
            .order_by(AgentReadinessSnapshot.computed_at.desc())
            .all()
        )
        latest: Dict[UUID, Dict] = {}
        for agent_id, overall_status, counts, run_number, computed_at in rows:
            # Rows arrive newest-first; keep only the first (latest) per agent.
            if agent_id in latest:
                continue
            tallies = _counts_from_mapping(counts)
            latest[agent_id] = {
                "overall_status": overall_status,
                "blocker_count": tallies["blockers"],
                "warning_count": tallies["warnings"],
                "info_count": tallies["info"],
                "run_number": run_number,
                "computed_at": computed_at.isoformat() if computed_at else None,
            }
        return latest

    # Cap the history scan so a pathological agent can't force a huge read.
    _MAX_RUNS_LIMIT = 100

    def list_runs(self, agent_id: str, *, limit: int = 20) -> ReadinessRunList:
        """List past **deep** readiness runs for one agent, newest first.

        Only deep runs (Test agent + publish gate) are surfaced — background
        shallow snapshots (editor loads, list-page impressions) are noise the
        user never asked to run and would swamp the dropdown. History spans all
        configs/versions of the agent; the UI badges cross-version rows.

        Metadata only — the heavy ``checks`` list is fetched per-run via
        ``get_run``. Org-scoped via ``_require_agent`` (404 for other orgs).
        """
        agent = self._require_agent(agent_id)
        capped = max(1, min(limit, self._MAX_RUNS_LIMIT))
        rows = (
            self.db.query(AgentReadinessEvent)
            .filter(
                AgentReadinessEvent.organization_id == self.org_id,
                AgentReadinessEvent.agent_id == agent.id,
                AgentReadinessEvent.depth == Depth.DEEP.value,
            )
            .order_by(AgentReadinessEvent.run_number.desc())
            .limit(capped)
            .all()
        )
        items: List[ReadinessRunListItem] = []
        for row in rows:
            try:
                items.append(_event_to_list_item(row))
            except (ValueError, KeyError):
                # One legacy/partial row with an out-of-enum status/counts must
                # not 500 the whole history list — skip it (with a traceback).
                logger.exception(
                    "[readiness] skipping unmappable run event run_number={}",
                    getattr(row, "run_number", None),
                )
        return ReadinessRunList(items=items)

    def get_run(self, agent_id: str, run_number: int) -> ReadinessReport:
        """Return the full stored report for one past run.

        Reuses ``_row_to_report`` — an event row carries the same
        ``ReadinessRowMixin`` columns as a snapshot, so the mapping is
        identical. Org-scoped; raises 404 for an unknown agent or run.
        """
        agent = self._require_agent(agent_id)
        row = (
            self.db.query(AgentReadinessEvent)
            .filter(
                AgentReadinessEvent.organization_id == self.org_id,
                AgentReadinessEvent.agent_id == agent.id,
                # Deep-only, mirroring ``list_runs`` — a shallow run_number that
                # never appears in the dropdown must 404, not return a report.
                AgentReadinessEvent.depth == Depth.DEEP.value,
                AgentReadinessEvent.run_number == run_number,
            )
            # Defensive: run_number is monotonic + concurrency-guarded so it's
            # effectively unique per agent, but there's no DB uniqueness
            # constraint — take the most recent row if two ever collide.
            .order_by(AgentReadinessEvent.computed_at.desc())
            .first()
        )
        if row is None:
            raise HTTPException(status_code=404, detail="Readiness run not found")
        try:
            return _row_to_report(row)
        except (ValueError, KeyError):
            # A stored row with an out-of-enum value can't be rendered — treat it
            # as unavailable rather than 500-ing (mirrors list_runs' skip). Log
            # the traceback so the corrupt row stays diagnosable.
            logger.exception(
                "[readiness] run {} unmappable for agent {}", run_number, agent_id
            )
            raise HTTPException(status_code=404, detail="Readiness run not found")

    async def gate_publish(
        self,
        agent_id: str,
        config_id: str,
        *,
        force_warnings: bool = False,
    ) -> None:
        """Publish gate: raise if the target config isn't safe to promote.

        Runs Deep (structural + live probes). BLOCKERs hard-fail; if only
        WARNINGs are present the caller must explicitly opt in with
        ``force_warnings=True``. Invalidates the in-memory Deep cache on
        success so the next readiness read reflects the freshly-published
        state (edit-based DB fast-path is naturally invalidated by the
        publish-time stamp change).
        """
        report = await self.check(
            agent_id,
            Depth.DEEP,
            config_id,
            trigger=TRIGGER_PUBLISH_GATE,
        )

        if report.overall_status == OverallStatus.NOT_READY:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail={
                    "reason": "readiness_blocked",
                    "message": (
                        f"{report.summary.get('blockers', 0)} blocker(s) prevent "
                        "publishing this version."
                    ),
                    "report": report.model_dump(),
                },
            )
        if (
            report.overall_status == OverallStatus.READY_WITH_WARNINGS
            and not force_warnings
        ):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail={
                    "reason": "readiness_warnings",
                    "message": (
                        f"{report.summary.get('warnings', 0)} warning(s) present. "
                        "Re-submit with force_warnings=true to publish anyway."
                    ),
                    "report": report.model_dump(),
                },
            )

        _deep_cache.invalidate(
            DeepCheckCache.key(self.org_id, agent_id, config_id)
        )

    # ── internals ──────────────────────────────────────────────────────────

    def _require_agent(self, agent_id: str) -> Agent:
        """Fetch an org-scoped agent or raise 404. Excludes soft-deleted rows."""
        try:
            aid = UUID(str(agent_id))
        except (ValueError, TypeError):
            raise HTTPException(status_code=400, detail="Invalid agent id")
        agent = (
            self.db.query(Agent)
            .filter(
                Agent.id == aid,
                Agent.organization_id == self.org_id,
                Agent.deleted_at.is_(None),
            )
            .first()
        )
        if agent is None:
            raise HTTPException(status_code=404, detail="Agent not found")
        return agent

    def _compute_stamp(self, agent: Agent):
        """Build the freshness key used by the snapshot fast-path.

        Combines the pipeline resolver's ``compute_agent_cache_version`` (which
        covers config + all provider/model/key/tool/KB/MCP/workflow deps) with
        readiness-only inputs the pipeline stamp ignores — currently just phone
        assignments, which live on the agent row rather than on any config.

        The pipeline helper is left untouched so runtime call caching stays
        unaffected; only readiness sees the augmented stamp. Never raises —
        any failure returns a ``None`` stamp so the fast-path misses cleanly
        and we recompute.
        """
        try:
            base_stamp, config = compute_agent_cache_version(
                self.db, agent, org_id=self.org_id
            )
        except Exception:  # noqa: BLE001
            logger.exception(
                "[readiness] dependency stamp computation failed for agent {}",
                agent.id,
            )
            return None, get_active_agent_config(self.db, agent)

        # A None base means the pipeline helper couldn't produce a stamp (agent
        # with no config); the fast-path must then miss — don't fabricate a
        # partial stamp that could accidentally match a stored one.
        if base_stamp is None:
            return None, config

        phone_stamp = self._phone_stamp(agent.id)
        return f"{base_stamp}|{phone_stamp}", config

    def _phone_stamp(self, agent_id: UUID) -> str:
        """Fingerprint the agent's phone assignments for the freshness key.

        Cheap single scalar-select — one round trip, no ORM materialisation.
        Format mirrors the other segments in ``compute_agent_cache_version``:
        ``phones:<count>:<max_updated_at_iso>`` (or ``:none`` when empty).
        """
        try:
            row = self.db.execute(
                select(
                    func.count(PhoneNumber.id),
                    func.max(PhoneNumber.updated_at),
                ).where(
                    PhoneNumber.agent_id == agent_id,
                    PhoneNumber.organization_id == self.org_id,
                )
            ).one()
            count, max_ts = row
            ts = max_ts.isoformat() if max_ts is not None else "none"
            return f"phones:{count or 0}:{ts}"
        except Exception:  # noqa: BLE001
            logger.exception(
                "[readiness] phone stamp computation failed for agent {}",
                agent_id,
            )
            # Deterministic sentinel — different from any real stamp so the
            # fast-path misses when we can't confidently reproduce state.
            return "phones:error"

    @staticmethod
    def _effective_config_id(
        requested_config_id: Optional[str], resolved_config
    ) -> Optional[UUID]:
        """The config id we actually check against: caller-supplied wins, else
        we resolve the active config (published or default)."""
        if requested_config_id:
            try:
                return UUID(str(requested_config_id))
            except (ValueError, TypeError):
                return None
        return resolved_config.id if resolved_config is not None else None

    def _read_fresh_snapshot(
        self,
        agent_id: UUID,
        config_id: Optional[UUID],
        depth: Depth,
        stamp: Optional[str],
    ) -> Optional[ReadinessReport]:
        """Return a stored snapshot iff its ``dependency_stamp`` matches the
        currently-live stamp. Missing stamp (either side) always misses.

        A **SHALLOW** request also accepts a **DEEP** snapshot: a deep run is a
        superset of shallow (all shallow checks + live provider probes), so for
        an unchanged config the most RECENT run governs the badge regardless of
        depth. Without this, reloading the page fires a shallow summary read
        that returns the stale shallow snapshot and resets the badge to "ready",
        discarding a "Run deep test" result (e.g. a live-probe blocker) the user
        just saw. A **DEEP** request (publish gate) still requires an actual deep
        snapshot. When both depths match the stamp, the newest by ``computed_at``
        wins — so a later shallow "Refresh" correctly supersedes an earlier deep
        run, and vice-versa."""
        if not stamp:
            return None
        q = self.db.query(AgentReadinessSnapshot).filter(
            AgentReadinessSnapshot.organization_id == self.org_id,
            AgentReadinessSnapshot.agent_id == agent_id,
            AgentReadinessSnapshot.dependency_stamp == stamp,
        )
        # Deep reads must not be satisfied by a shallow snapshot; shallow reads
        # accept either depth (see docstring).
        if depth == Depth.DEEP:
            q = q.filter(AgentReadinessSnapshot.depth == Depth.DEEP.value)
        # SQLAlchemy translates ``== None`` to ``IS NULL`` — needed so a
        # brand-new-agent snapshot (config_id NULL) can still fast-path.
        if config_id is None:
            q = q.filter(AgentReadinessSnapshot.config_id.is_(None))
        else:
            q = q.filter(AgentReadinessSnapshot.config_id == config_id)
        row = q.order_by(AgentReadinessSnapshot.computed_at.desc()).first()
        if row is None:
            return None
        if depth == Depth.DEEP and _deep_snapshot_expired(row):
            return None
        return _row_to_report(row)

    async def _run_and_persist(
        self,
        *,
        agent: Agent,
        config_id: Optional[str],
        depth: Depth,
        trigger: str,
        dependency_stamp: Optional[str],
    ) -> ReadinessReport:
        """Compute a fresh report, persist it, and return it. Timing is
        captured for the ``duration_ms`` column, and ``started_at`` /
        ``run_number`` are stamped at persist time so both snapshot and event
        rows carry identical provenance."""
        started_wall = datetime.now(timezone.utc)
        started = time.monotonic()
        report = await self._run(agent, config_id, depth)
        duration_ms = int((time.monotonic() - started) * 1000)

        # Stamp provenance on the in-memory report so the API response carries
        # the same fields the stored row will, without a follow-up DB read.
        run_number = self._next_run_number(agent.id)
        report.started_at = started_wall.isoformat()
        report.run_number = run_number

        # Persistence is fire-and-forget from the caller's perspective — a
        # persistence failure logs a warning inside ``.record()`` but never
        # raises. We still guard the stamp being None (fast-path miss reason).
        if dependency_stamp:
            ReadinessPersistence(self.db, self.org_id).record(
                report,
                trigger=trigger,
                triggered_by_user_id=self.user_id,
                dependency_stamp=dependency_stamp,
                duration_ms=duration_ms,
                started_at=started_wall,
                run_number=run_number,
            )
        return report

    def _next_run_number(self, agent_id: UUID) -> int:
        """Return the next run-number for this agent.

        Reads the highest ``run_number`` stored in the append-only event log
        and adds one. Rate limiter + coalesce cache prevent concurrent
        readiness runs for the same agent, so ``MAX + 1`` is race-free in
        practice. Falls back to 1 for the very first run.
        """
        current = self.db.execute(
            select(func.max(AgentReadinessEvent.run_number))
            .where(AgentReadinessEvent.agent_id == agent_id)
        ).scalar()
        return (current or 0) + 1

    async def _run(
        self,
        agent: Agent,
        config_id: Optional[str],
        depth: Depth,
        *,
        deep_categories: Optional[Set[Category]] = None,
    ) -> ReadinessReport:
        builder = ContextBuilder(self.db, self.org_id)
        config = builder.resolve_config(agent, config_id=config_id)
        ctx = builder.build(agent, config, depth)
        return await Runner().run(ctx, deep_categories=deep_categories)


# ── module-private helpers ─────────────────────────────────────────────────


def _deep_snapshot_expired(row: AgentReadinessSnapshot) -> bool:
    """True when a stored Deep snapshot is older than the wall-clock TTL.

    A missing ``computed_at`` is defensively treated as expired — that would
    only happen for hand-inserted rows, and we'd rather re-run than serve a
    snapshot with no provenance.
    """
    if row.computed_at is None:
        return True
    age = datetime.now(timezone.utc) - row.computed_at
    return age > timedelta(seconds=_DEEP_SNAPSHOT_TTL_SECONDS)


def _counts_from_mapping(stored: Optional[Dict]) -> Dict[str, int]:
    """Extract the five severity/status tallies from a raw ``counts`` mapping,
    defaulting any missing key to 0 so a row predating a new count key still
    renders. The single source of truth for the fallback set — used by the row
    mappers (via ``_counts_from_row``) and by the batch list-badge lookup
    (``latest_for_agents``), which projects the ``counts`` column directly."""
    stored = stored or {}
    return {
        "blockers": int(stored.get("blockers", 0)),
        "warnings": int(stored.get("warnings", 0)),
        "info": int(stored.get("info", 0)),
        "passed": int(stored.get("passed", 0)),
        "skipped": int(stored.get("skipped", 0)),
    }


def _counts_from_row(row: Union[AgentReadinessSnapshot, AgentReadinessEvent]) -> Dict[str, int]:
    """Tallies for a full row — delegates to ``_counts_from_mapping``."""
    return _counts_from_mapping(row.counts)


def _row_to_report(row: Union[AgentReadinessSnapshot, AgentReadinessEvent]) -> ReadinessReport:
    """Turn a stored readiness row back into a ``ReadinessReport``.

    Accepts either an ``AgentReadinessSnapshot`` (latest state) or an
    ``AgentReadinessEvent`` (history) — both carry the identical
    ``ReadinessRowMixin`` columns, so one mapper serves both. Both JSONB
    columns (``counts`` and ``checks``) already carry the exact wire shape so
    there's no field-by-field mapping — just hand them to the Pydantic
    models. Any missing count key falls back to 0 so an older row predating a
    new severity level still renders."""
    checks: List[CheckResult] = [
        CheckResult.model_validate(c) for c in (row.checks or [])
    ]
    return ReadinessReport(
        agent_id=str(row.agent_id),
        config_id=str(row.config_id) if row.config_id else None,
        depth=Depth(row.depth),
        overall_status=OverallStatus(row.overall_status),
        summary=_counts_from_row(row),
        checks=checks,
        generated_at=row.computed_at.isoformat() if row.computed_at else "",
        started_at=row.started_at.isoformat() if row.started_at else None,
        run_number=row.run_number,
    )


def _event_to_list_item(row: AgentReadinessEvent) -> ReadinessRunListItem:
    """Flatten a history event row into a lightweight dropdown item.

    Reuses ``_counts_from_row`` for the same defensive tallies as
    ``_row_to_report`` but drops the heavy ``checks`` list — the dropdown only
    needs summary tallies and provenance; full checks load lazily via
    ``get_run``."""
    counts = _counts_from_row(row)
    return ReadinessRunListItem(
        run_number=row.run_number,
        depth=Depth(row.depth),
        overall_status=OverallStatus(row.overall_status),
        config_id=str(row.config_id) if row.config_id else None,
        trigger=row.trigger,
        blocker_count=counts["blockers"],
        warning_count=counts["warnings"],
        info_count=counts["info"],
        passed_count=counts["passed"],
        skipped_count=counts["skipped"],
        started_at=row.started_at.isoformat() if row.started_at else None,
        computed_at=row.computed_at.isoformat() if row.computed_at else "",
        duration_ms=row.duration_ms,
    )
